Userland/Libraries/LibWeb/CSS/Parser/Tokenizer.cpp

```diff
@@ -469,7 +470,73 @@ CSSNumber Tokenizer::consume_a_number()
         }
     }
 
-    return { repr.to_string(), type };
+    return { repr.to_string(), convert_a_string_to_a_number(repr.string_view()), type };
+}
+
+// https://www.w3.org/TR/css-syntax-3/#convert-string-to-number
+double Tokenizer::convert_a_string_to_a_number(StringView string)
+{
+    auto code_point_at = [&](size_t index) -> u32 {
+        if (index < string.length())
+            return string[index];
+        return TOKENIZER_EOF;
+    };
+
+    // Divide the string into seven components, in order from left to right:
+    size_t position = 0;
+
+    // 1. A sign: a single U+002B PLUS SIGN (+) or U+002D HYPHEN-MINUS (-), or the empty string.
+    //    Let s be the number -1 if the sign is U+002D HYPHEN-MINUS (-); otherwise, let s be the number 1.
+    int s = 1;
+    if (is_plus_sign(code_point_at(position)) || is_hyphen_minus(code_point_at(position))) {
+        s = is_hyphen_minus(code_point_at(position)) ? -1 : 1;
+        position++;
+    }
+
+    // 2. An integer part: zero or more digits.
+    //    If there is at least one digit, let i be the number formed by interpreting the digits as a base-10 integer; otherwise, let i be the number 0.
+    int i = 0;
+    while (is_ascii_digit(code_point_at(position))) {
+        i = (i * 10) + (code_point_at(position) - '0');
+        position++;
+    }
+
+    // 3. A decimal point: a single U+002E FULL STOP (.), or the empty string.
+    if (is_full_stop(code_point_at(position)))
+        position++;
+
+    // 4. A fractional part: zero or more digits.
+    //    If there is at least one digit, let f be the number formed by interpreting the digits as a base-10 integer and d be the number of digits; otherwise, let f and d be the number 0.
+    int f = 0;
+    int d = 0;
+    while (is_ascii_digit(code_point_at(position))) {
+        f = (f * 10) + (code_point_at(position) - '0');
+        position++;
+        d++;
+    }
+
+    // 5. An exponent indicator: a single U+0045 LATIN CAPITAL LETTER E (E) or U+0065 LATIN SMALL LETTER E (e), or the empty string.
+    if (is_e(code_point_at(position)) || is_E(code_point_at(position)))
+        position++;
+
+    // 6. An exponent sign: a single U+002B PLUS SIGN (+) or U+002D HYPHEN-MINUS (-), or the empty string.
+    //    Let t be the number -1 if the sign is U+002D HYPHEN-MINUS (-); otherwise, let t be the number 1.
+    int t = 1;
+    if (is_plus_sign(code_point_at(position)) || is_hyphen_minus(code_point_at(position))) {
+        t = is_hyphen_minus(code_point_at(position)) ? -1 : 1;
+        position++;
+    }
+
+    // 7. An exponent: zero or more digits.
+    //    If there is at least one digit, let e be the number formed by interpreting the digits as a base-10 integer; otherwise, let e be the number 0.
+    int e = 0;
+    while (is_ascii_digit(code_point_at(position))) {
+        e = (e * 10) + (code_point_at(position) - '0');
+        position++;
+    }
+
```

Comment:
The spec doesn't care if there's extra garbage at the end of the string?

Reply:
Hmmm. The way the spec defines it, we've already made sure that the string is a valid number definition before we call this, so there never should be anything left over:

> Note: This algorithm does not do any verification to ensure that the string contains only a number. Ensure that the string contains only a valid CSS number before calling this algorithm.

 I'll add a verify at the end though, good catch.